Source/WebCore/Modules/permissions/PermissionStatus.cpp

```diff
@@ -30,49 +30,87 @@
+{
+    auto status = adoptRef(*new PermissionStatus(context, state, descriptor, source, page));
+    if (source != PermissionQuerySource::Window)
+        status->createAssociatedMainThreadPermissionObserver(context);
     status->suspendIfNeeded();
     return status;
 }
 
-PermissionStatus::PermissionStatus(ScriptExecutionContext& context, PermissionState state, const PermissionDescriptor& descriptor)
+PermissionStatus::PermissionStatus(ScriptExecutionContext& context, PermissionState state, const PermissionDescriptor& descriptor, PermissionQuerySource source, Page* page)
     : ActiveDOMObject(&context)
     , m_state(state)
     , m_descriptor(descriptor)
+    , m_source(source)
+    , m_page(page)
 {
     auto* origin = context.securityOrigin();
     auto originData = origin ? origin->data() : SecurityOriginData { };
     m_origin = ClientOrigin { context.topOrigin().data(), WTFMove(originData) };
     
-    // FIXME: Add support for workers. Currently, this only works for Window objects.
-    if (isMainThread())
+    if (source == PermissionQuerySource::Window)
         PermissionController::shared().addObserver(*this);
 }
 
+void PermissionStatus::createAssociatedMainThreadPermissionObserver(ScriptExecutionContext& context)
+{
+    auto& workerGlobalScope = downcast<WorkerGlobalScope>(context);
+    callOnMainThread([this, protectedThis = Ref { *this }, contextIdentifier = workerGlobalScope.identifier(), origin = m_origin.isolatedCopy()]() mutable {
+        auto mainThreadPermissionObserver = makeUnique<MainThreadPermissionObserver>(this, contextIdentifier, m_state, m_descriptor, m_source, m_page.get(), origin);
+        auto mainThreadPermissionObserverIdentifier = mainThreadPermissionObserver->identifier();
+
+        allMainThreadPermissionObservers().add(mainThreadPermissionObserverIdentifier, WTFMove(mainThreadPermissionObserver));
+
+        ScriptExecutionContext::postTaskTo(contextIdentifier, [protectedThis = WTFMove(protectedThis), identifier = mainThreadPermissionObserverIdentifier](auto&) {
+            protectedThis->m_mainThreadPermissionObserverIdentifier = identifier;
+        });
+    });
+}
+
 PermissionStatus::~PermissionStatus()
 {
-    // FIXME: Add support for workers. Currently, this only works for Window objects.
-    if (isMainThread())
-        PermissionController::shared().removeObserver(*this);
+    if (!m_mainThreadPermissionObserverIdentifier)
+        return;
+
+    callOnMainThread([identifier = *m_mainThreadPermissionObserverIdentifier]() {
+        allMainThreadPermissionObservers().remove(identifier);
+    });
 }
 
 void PermissionStatus::stateChanged(PermissionState newState)
 {
     if (m_state == newState)
         return;
 
-    if (auto* document = dynamicDowncast<Document>(scriptExecutionContext()); !document->isFullyActive())
+    auto* document = dynamicDowncast<Document>(scriptExecutionContext());
+    if (document && !document->isFullyActive())
```

Comment:
[Spec](https://w3c.github.io/permissions/#onchange-attribute) says:
`If this's relevant global object is a Window object, ... If document is null or document is not fully active, terminate this algorithm.`
Shouldn't we also return if scriptExecutionContext() is null?